### ISSUE TITLE
Motorstorm resolution scale mudtracks causes slowdown(#6774)

### ISSUE THREAD

**Issue by JimScript** [RPCS3/rpcs3#11898]
## Quick summary
After #6774, vehicles diving in the mud at higher resolution causes a serious GPU penalty in Motorstorm(BCUS98137).

## Details
Before #6774, Driving around the track gave a gave a consistent gpu performance, but now anytime any vehicle drives over mud it spikes the gpu usage exponentially with the resolution scale. It is also as of now worse(or at the very least more stark) then when the regression first occurred. It is also weird in that the performance penalty only occurs when the game is unpaused, despite the fact you can still rotate the camera when paused.

#### 1. Please provide _exact_ build (or commit) information that introduced the regression you're reporting.
#6774 is when the regression started

#### 2. Please attach TWO logs:
* One for the build with regression:[RPCS3.log](https://github.com/RPCS3/rpcs3/files/8589745/RPCS3.log)
* One for the build that works as expected:[RPCS3.log](https://github.com/RPCS3/rpcs3/files/8589721/RPCS3.log)
* Additional log for current RPCS3:[RPCS3.log](https://github.com/RPCS3/rpcs3/files/8589783/RPCS3.log)

#### 4. Please attach screenshots of your problem.
These are all at 300% scale(4K)

Before:
![FPS_ 0 00 _ Vulkan _ 0 0 7-8909 _ MotorStorm™  BCUS98137  4_29_2022 2_18_05 AM](https://user-images.githubusercontent.com/66685584/165919795-b5b9688e-4b85-4809-9343-038cb7e85d30.png)

After:
![FPS_ 0 00 _ Vulkan _ 0 0 7-8920 _ MotorStorm™  BCUS98137  4_29_2022 2_21_15 AM](https://user-images.githubusercontent.com/66685584/165920066-4b771646-2026-4060-97c1-a0ae02d07190.png)

Current unpaused:
![FPS_ 0 00 _ Vulkan _ 0 0 21-13532 _ MotorStorm™  BCUS98137  4_29_2022 2_28_05 AM](https://user-images.githubusercontent.com/66685584/165920471-6f85b776-1aa0-42fa-888e-5028dfaa2da6.png)

Current paused:
![FPS_ 0 00 _ Vulkan _ 0 0 21-13532 _ MotorStorm™  BCUS98137  4_29_2022 2_28_11 AM](https://user-images.githubusercontent.com/66685584/165920200-d95c1c2a-d76e-44eb-bdf8-cb256ad42c94.png)

#### 6. Please provide your system configuration:
- SYS: RPCS3 v0.0.21-13532-8316469c Alpha | master
- SYS: AMD Ryzen 5 3600 6-Core Processor | 12 Threads | 15.95 GiB RAM | TSC: 3.600GHz | AVX+ | FMA3
- SYS: Operating system: Windows, Major: 10, Minor: 0, Build: 19043, Service Pack: none, Compatibility mode: 0
- RSX: Found vulkan-compatible GPU: 'NVIDIA GeForce RTX 2070 SUPER' running on driver 511.23.0.0
- SYS: Firmware version: 4.88

**Comment by hoholee12** [comment]
Use 'Disable Dynamic Resolution Scaling' from patch manager.

**Comment by JimScript** [comment]
None of the patches seem to help. When driving normally at 200% resolution scale, Msi afterburner reports ~40% gpu usage on my RTX 2070 super, but when _any_ vehicle(even ones offscreen) goes through the mud it jumps to ~80% gpu usage. It goes back to ~40% when you pause or all of the vehicles get off the mud.

**Comment by JimScript** [comment]
With PSONE providing the 3.1 update, I can now finally test the tracks in a more controlled environment with time attack.
2k with the unlock fps and disable dynamic resolution patches:

https://user-images.githubusercontent.com/66685584/174453118-e69a9caf-5f4f-49c0-836a-b5e95362b157.mp4

**Comment by kd-11** [comment]
Retest. There was some recent logic to discard out-of-range slices, though if the GPU is getting hit that hard, it is likely doing real work. I'm also removing the regression label, as the drop in performance is the result of doing actual intentional work.
We do have optimizations to try and keep the GPU from getting spammed too much with update requests, I wonder why they don't kick in.

**Comment by JimScript** [comment]
Well I went back to this exact spot, and it seems to run just the same as before. Maybe the update request are just huge instead of frequent. It feels like anytime anything moves in the mud, it just continuously re-updates the entire level's mud buffer, which is too big to be cut off by the resolution scale threshold. Maybe it is suppose to run this hard, but even though my RTX 2070s can run this easily enough at original settings, my steam deck is maxed out on the GPU.

**Comment by JimScript** [comment]
OpenGL doesn't seem to draw the mudtracks, but the weird thing is they persist from a savestate until they are driven on again:

https://user-images.githubusercontent.com/66685584/194425203-fd00a1c3-dd9f-401c-88a9-a9b97a41061e.mp4

So it _does_ completely overwrite the whole level's mud every single tick, along with the fact it doesn't work in OpenGL.

**Comment by JimScript** [comment]
I did a little RenderDoc and found that the reason why the mud tracks reset after savestate is because they are saved in a different way then they are rendered. The game at its base resolution renders the mudtracks in a _4096x4096_ R8_UNORM texture with 9 mips, so when you bump it up to 2k it becomes 8192x8192(I don't even want to think how big it gets a 4k). But savestates always saves it to 4096x4096 with 13 mips regardless of what it was before, and it gets thrown out as soon as it has to re-render the mud. It kinda makes me want a Resolution Scale Maximum as well as the Resolution Scale Minimum we have now, make the threshold 2 way.

**Comment by JimScript** [comment]
Well, here is what the buffer looks like a base resolution:
![MotorstormMudMap](https://user-images.githubusercontent.com/66685584/195673749-1c2649e5-f722-455e-929e-e41b93ac41de.png)

**Comment by JimScript** [comment]
So, why does the resolution scale minimum max out at 1024 with no resolution scale maximum option? I get this could be a UX thing, but you can't even set it higher in a custom configuration. So if I say wanted to lower buffers bigger than 2048 by 50% to save GPU performance, I couldn't. I heard that 4096 was the maximum texture/buffer size the RSX could handle, so why doesn't it go over 1024? How hard(if even possible) would it be to change the 1024 min to 4096 min + 4096 max, even if it is not in the UI? I guess there was never a need until now.

**Comment by kd-11** [comment]
Because th

**Comment by JimScript** [comment]
> Can confirm this issue still presents in latest build 0.0.28-15214 as of June 2023. Game runs great at 720p and up to 1080p, but 1440p and up require powerful hardware to handle this scaling issue. Would be useful to test if this is indeed the totality of the performance impact if it were possible to toggle the mud tracks off, or if there is another post-processing effect also impacting scaling.

720p still requires moderately powerful hardware since the Steam Deck's GPU is basically maxed out by this effect(near 100% at 1600Mhz in time trial). And the only way I could see the mud tracks turned off is if someone made a patch for it, I tried a bit but found that I was severely out of my depth with patching. I was trying to find where the 4096x4096 resolution of the mud tracks was defined so that I could lower it to around 1024 or something, that way it would be smaller than normal even at 4k.

**Comment by JimScript** [comment]
I found the renderdoc compatibility mode to see what the mud-maps signature is; turns out it uses a 4096x4096 texture that already had detail to it, _but_ when anything goes over it over it throws out that original mud-map and just writes to a blank "Temp view" that scales to the resolution. It should be drawing _onto_ that texture but is instead replacing it, so this is more than just a performance issue, there are also missing details from this unused base mud-map.

**Comment by kd-11** [comment]
A few things to note:
1. It doesn't matter if the mud map is replaced, as long as the visuals are ok on the output.
2. If it scales to resolution that means the map is not a texture but a render target.
3. Temp views have several use cases, including mipmap gathering or texture atlas. We never use a temp view as-is, so check what we're actually doing with it if 1 is true. It should be a destination for a copy operation at some point before the call.

Also, let's focus on the actual bug, don't log implementation details as bugs unless it's proven. It can derail investigations. For example, none of this has anything to do with the performance hit by itself unless it can be shown that the temp copy is unnecessary (e.g it is always the same size, so no scaling is required)

**Comment by JimScript** [comment]
I suppose I should've showed rather than told what the problem is then.
Here is what the mud-map looks like when I loaded it in:
![MotorstormMudMapOGgray](https://github.com/RPCS3/rpcs3/assets/66685584/d6578cf6-28aa-4af0-9e8d-51637430a97a)


And here is what it looks like when I drove around it for a bit(it's 4x the resolution because of the 200% scaling):
![MotorstormMudMapBrokegray](https://github.com/RPCS3/rpcs3/assets/66685584/7f2d2411-dd6e-42ae-8da9-083684c617cd)


And this is what I think it's suppose to look like:
![MotorstormMudMapFixedgray](https://github.com/RPCS3/rpcs3/assets/66685584/847b3c08-1679-44c7-af60-d8ee82566645)

The last video I posted on this issue showed the previous mudtracks just popping out of existence as soon as you drive on the track, and this can be proven by loading into a time trial on the grizzly and going backwards on the track until you reach a ramp. On every race as soon as your tires touch the mud a bunch detail on the mud is lost for the entire race until you restart, and it's not like this on a real PS3.

**Comment by kd-11** [comment]
Ok, I think that's a separate issue, though it is unusual for the data to go "missing" unless RCB is not enabled. If you're looking into both of them you can keep them together like this I suppose, but since they require different PRs / code changes you may be better off creating a separate ticket for the visuals issue.

On that note, I'm not convinced the performance issue is actionable in a reasonable manner, especially as it comes from the massively increased resolution of some internal buffers when changing the scaling mode. I recently (June/July) added virtual subimages to avoid copying around massive image slices for the gather operations. Unfortunately this is not possible for mipmap slices without implementing a full virtual texturing pipeline (including partial image filtering steps) which would be much slower on older hardware than just copying the data a few times every frame. Has performance changed recently? Is it still slow? The bottleneck would be GPU memory bandwidth (esp bus width) which can also explain the hardware-sensitive nature of the issue.

**Comment by kd-11** [comment]
I now realize I never fully explained why max res scaling breaks logic. It's fairly simple, we keep things at a ratio to each other and try to exclude some small buffers that are usually not used for rendering graphics. The threshold isn't meant to make graphical changes but to try and filter out compute resources (PS3 is DX9 hw, so compute resources are emulated as small square textures)
Now, in many places, such as mipmap generation, atlas generation, etc we have to stitch together a bunch of chopped up surfaces and we rely on the sizes at least being consistent relative to each other for the generated output to be correct. This is also why resolution threshold can still sometimes break things. However, in the case of small images, we're very unlikely to introduce problems compared to having a cap, since it is typical for shadowmaps and other projected effects (sunshafts, etc) to use large buffers. The typical PS3 shadowmap is 1024x1024 to 2048x2048. Some may even go up to 4096 per side at 720p just because shadows demand so much resolution. A per-aspect clamp also wouldn't work since games switch between aspects (basically GPU reinterpret cast) all the time.

**Comment by JimScript** [comment]
> Ok, I think that's a separate issue, though it is unusual for the data to go "missing" unless RCB is not enabled. If you're looking into both of them you can keep them together like this I suppose, but since they require different PRs / code changes you may be better off creating a separate ticket for the visuals issue.
> 
> On that note, I'm not convinced the performance issue is actionable in a reasonable manner, especially as it comes from the massively increased resolution of some internal buffers when changing the scaling mode. I recently (June/July) added virtual subimages to avoid copying around massive image slices for the gather operations. Unfortunately this is not possible for mipmap slices without implementing a full virtual texturing pipeline (including partial image filtering steps) which would be much slower on older hardware than just copying the data a few times every frame. Has performance changed recently? Is it still slow? The bottleneck would be GPU memory bandwidth (esp bus width) which can also explain the hardware-sensitive nature of the issue. Of course a patch wouldn't care about this kind of thing. There are other hacky alternatives to this, but I'm not comfortable adding code that would likely break other games with little benefit.

Well the RCB did the trick, so I'll stop with the whole "original mud-map" thing. And you're probably right, this probably isn't an actionable problem that needs attention, it's probably just a fact of life that the emulator will have to deal with these huge buffers. I guess this game will just need a strong GPU to run then. I think I should just close this then, I'm sorry for wasting your time.

**Comment by JimScript** [comment]
Actually I'll let you decide if it's still worth pursuing, as there still are slowdowns from higher resolution scaling.

**Comment by kd-11** [comment]
If performance at stock and ~1080p is ok then we can close it. What's left is just part of general optimization work as there is nothing technically wrong or incorrect with what the game is doing.
